src/infra/core/src/repos/dataset_repository_s3.rs

```diff
@@ -415,3 +564,27 @@ impl DatasetRepository for DatasetRepositoryS3 {
 }
 
 /////////////////////////////////////////////////////////////////////////////////////////
+
+pub struct S3RegistryCache {
+    state: Arc<Mutex<State>>,
```

Comment:
This is far from elegant but I think we will factor it out into a proper registry component in future.
I'd rather save this refactoring until #342.